File: udon_decompiler/analysis/transform/ir_utils.py
from __future__ import annotations

from collections import deque
from typing import Dict, Iterator, List, Optional, Set

from udon_decompiler.analysis.ir.nodes import (
    IRBlock,
    IRBlockContainer,
    IRFunction,
    IRHighLevelDoWhile,
    IRHighLevelSwitch,
    IRHighLevelWhile,
    IRIf,
    IRJump,
    IRLeave,
    IRReturn,
    IRStatement,
    IRSwitch,
)


def is_terminator(statement: IRStatement) -> bool:
    return isinstance(statement, (IRIf, IRJump, IRLeave, IRReturn, IRSwitch))


def get_block_terminator(block: IRBlock) -> Optional[IRStatement]:
    if not block.statements:
        return None
    last = block.statements[-1]
    if is_terminator(last):
        return last
    return None


def iter_block_targets(statement: IRStatement) -> Iterator[IRBlock]:
    if isinstance(statement, IRJump):
        yield statement.target
        return

    if isinstance(statement, IRIf):
        yield from iter_block_targets(statement.true_statement)
        if statement.false_statement is not None:
            yield from iter_block_targets(statement.false_statement)
        return

    if isinstance(statement, IRSwitch):
        for block in statement.cases.values():
            yield block
        if statement.default_target is not None:
            yield statement.default_target


def _build_successor_map(container: IRBlockContainer) -> Dict[IRBlock, List[IRBlock]]:
    block_set = set(container.blocks)
    successors: Dict[IRBlock, List[IRBlock]] = {block: [] for block in container.blocks}

    for block in container.blocks:
        terminator = get_block_terminator(block)
        if terminator is None:
            continue
        for target in iter_block_targets(terminator):
            if target in block_set:
                successors[block].append(target)

    return successors


def reachable_blocks(container: IRBlockContainer) -> List[IRBlock]:
    if container.entry_block is None:
        return []

    successors = _build_successor_map(container)
    entry = container.entry_block

    visited: Set[IRBlock] = set()
    queue: deque[IRBlock] = deque([entry])
    while queue:
        block = queue.popleft()
        if block in visited:
            continue
        visited.add(block)
        for succ in successors.get(block, []):
            if succ not in visited:
                queue.append(succ)
    return [b for b in container.blocks if b in visited]


def iter_block_containers(function: IRFunction) -> Iterator[IRBlockContainer]:
    visited: set[int] = set()
    queue: deque[IRBlockContainer] = deque([function.body])

    while queue:
        container = queue.popleft()
        container_id = id(container)
        if container_id in visited:
            continue
        visited.add(container_id)
        yield container

        for block in container.blocks:
            for statement in block.statements:
                queue.extend(_iter_nested_containers(statement))


def _iter_nested_containers(statement: IRStatement) -> Iterator[IRBlockContainer]:
    if isinstance(statement, IRBlockContainer):
        yield statement
        for block in statement.blocks:
            for nested_statement in block.statements:
                yield from _iter_nested_containers(nested_statement)
        return

    if isinstance(statement, IRBlock):
        for nested_statement in statement.statements:
            yield from _iter_nested_containers(nested_statement)
        return

    if isinstance(statement, IRIf):
        yield from _iter_nested_containers(statement.true_statement)
        if statement.false_statement is not None:
            yield from _iter_nested_containers(statement.false_statement)
        return

    if isinstance(statement, IRHighLevelSwitch):
        for section in statement.sections:
            yield from _iter_nested_containers(section.body)
        return

    if isinstance(statement, (IRHighLevelWhile, IRHighLevelDoWhile)):
        yield from _iter_nested_containers(statement.body)

File: src/decompiler/transform/ir_utils.rs
use std::collections::{HashSet, VecDeque};

use crate::decompiler::ir::{
    IrBlock, IrBlockContainer, IrFunction, IrHighLevelDoWhile, IrHighLevelSwitch, IrHighLevelWhile,
    IrIf, IrJump, IrStatement, IrSwitch,
};

pub fn get_block_terminator(block: &IrBlock) -> Option<&IrStatement> {
    let last = block.statements.last()?;
    if last.is_terminator() {
        Some(last)
    } else {
        None
    }
}

pub fn iter_block_targets(statement: &IrStatement) -> Vec<u32> {
    match statement {
        IrStatement::Jump(IrJump { target_address }) => vec![*target_address],
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            let mut out = iter_block_targets(true_statement);
            if let Some(false_statement) = false_statement.as_ref() {
                out.extend(iter_block_targets(false_statement));
            }
            out
        }
        IrStatement::Switch(IrSwitch {
            cases,
            default_target,
            ..
        }) => {
            let mut out = cases.values().copied().collect::<Vec<_>>();
            if let Some(default_target) = default_target {
                out.push(*default_target);
            }
            out
        }
        _ => Vec::new(),
    }
}

pub fn iter_block_containers(function: &IrFunction) -> Vec<u32> {
    let mut out = Vec::<u32>::new();
    let mut visited = HashSet::<u32>::new();
    let mut queue = VecDeque::<u32>::new();
    queue.push_back(function.body.id);

    while let Some(container_id) = queue.pop_front() {
        if !visited.insert(container_id) {
            continue;
        }
        out.push(container_id);

        let mut nested = Vec::<u32>::new();
        if function.body.id == container_id {
            collect_nested_containers_from_container(&function.body, &mut nested);
        } else {
            // caller traverses nested containers by repeatedly resolving ids;
            // this helper only knows root container directly.
            // nested ids for non-root are discovered when that container is visited.
            if let Some(container) = find_container(function, container_id) {
                collect_nested_containers_from_container(container, &mut nested);
            }
        }
        for id in nested {
            queue.push_back(id);
        }
    }

    out
}

pub fn find_container(function: &IrFunction, container_id: u32) -> Option<&IrBlockContainer> {
    find_container_in_container(&function.body, container_id)
}

pub fn find_container_mut(
    container: &mut IrBlockContainer,
    container_id: u32,
) -> Option<&mut IrBlockContainer> {
    if container.id == container_id {
        return Some(container);
    }

    for block in &mut container.blocks {
        for statement in &mut block.statements {
            if let Some(found) = find_container_in_statement_mut(statement, container_id) {
                return Some(found);
            }
        }
    }
    None
}

fn collect_nested_containers_from_container(container: &IrBlockContainer, out: &mut Vec<u32>) {
    for block in &container.blocks {
        for statement in &block.statements {
            collect_nested_containers_from_statement(statement, out);
        }
    }
}

fn collect_nested_containers_from_statement(statement: &IrStatement, out: &mut Vec<u32>) {
    match statement {
        IrStatement::BlockContainer(container) => {
            out.push(container.id);
            collect_nested_containers_from_container(container, out);
        }
        IrStatement::Block(block) => {
            for nested in &block.statements {
                collect_nested_containers_from_statement(nested, out);
            }
        }
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            collect_nested_containers_from_statement(true_statement, out);
            if let Some(false_statement) = false_statement.as_ref() {
                collect_nested_containers_from_statement(false_statement, out);
            }
        }
        IrStatement::HighLevelSwitch(IrHighLevelSwitch { sections, .. }) => {
            for section in sections {
                for statement in &section.body.statements {
                    collect_nested_containers_from_statement(statement, out);
                }
            }
        }
        IrStatement::HighLevelWhile(IrHighLevelWhile { body, .. }) => {
            out.push(body.id);
            collect_nested_containers_from_container(body, out);
        }
        IrStatement::HighLevelDoWhile(IrHighLevelDoWhile { body, .. }) => {
            out.push(body.id);
            collect_nested_containers_from_container(body, out);
        }
        _ => {}
    }
}

fn find_container_in_container(
    container: &IrBlockContainer,
    container_id: u32,
) -> Option<&IrBlockContainer> {
    if container.id == container_id {
        return Some(container);
    }

    for block in &container.blocks {
        for statement in &block.statements {
            if let Some(found) = find_container_in_statement(statement, container_id) {
                return Some(found);
            }
        }
    }
    None
}

fn find_container_in_statement(
    statement: &IrStatement,
    container_id: u32,
) -> Option<&IrBlockContainer> {
    match statement {
        IrStatement::BlockContainer(container) => {
            find_container_in_container(container, container_id)
        }
        IrStatement::Block(block) => {
            for nested in &block.statements {
                if let Some(found) = find_container_in_statement(nested, container_id) {
                    return Some(found);
                }
            }
            None
        }
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            if let Some(found) = find_container_in_statement(true_statement, container_id) {
                return Some(found);
            }
            if let Some(false_statement) = false_statement.as_ref() {
                return find_container_in_statement(false_statement, container_id);
            }
            None
        }
        IrStatement::HighLevelWhile(IrHighLevelWhile { body, .. }) => {
            find_container_in_container(body, container_id)
        }
        IrStatement::HighLevelDoWhile(IrHighLevelDoWhile { body, .. }) => {
            find_container_in_container(body, container_id)
        }
        _ => None,
    }
}

fn find_container_in_statement_mut(
    statement: &mut IrStatement,
    container_id: u32,
) -> Option<&mut IrBlockContainer> {
    match statement {
        IrStatement::BlockContainer(container) => find_container_mut(container, container_id),
        IrStatement::Block(block) => {
            for nested in &mut block.statements {
                if let Some(found) = find_container_in_statement_mut(nested, container_id) {
                    return Some(found);
                }
            }
            None
        }
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            if let Some(found) = find_container_in_statement_mut(true_statement, container_id) {
                return Some(found);
            }
            if let Some(false_statement) = false_statement.as_mut()
                && let Some(found) = find_container_in_statement_mut(false_statement, container_id)
            {
                return Some(found);
            }
            None
        }
        IrStatement::HighLevelWhile(IrHighLevelWhile { body, .. }) => {
            find_container_mut(body, container_id)
        }
        IrStatement::HighLevelDoWhile(IrHighLevelDoWhile { body, .. }) => {
            find_container_mut(body, container_id)
        }
        _ => None,
    }
}
